client/src/includes/userbar.ts

```diff
@@ -0,0 +1,458 @@
+
+    // Make sure userbar is hidden when navigating back
+    window.addEventListener('pageshow', hideUserbar, false);
+
+    // Handle keyboard events on the trigger
+    userbar.addEventListener('keydown', handleTriggerKeyDown);
+    list.addEventListener('focusout', handleFocusChange);
+
+    // On initialisation, all menu items should be disabled for roving tab index
+    resetItemsTabIndex();
+
+    document.addEventListener('DOMContentLoaded', async () => {
+      await this.initialiseAxe();
+    });
+  }
+
+  /*
+  Integrating Axe accessibility checker to improve ATAG compliance, adapted for content authors to identify and fix accessibility issues.
+  Scans loaded page for errors with 3 initial rules ('empty-heading', 'p-as-heading', 'heading-order') and outputs the results in GUI.
+  See documentation: https://github.com/dequelabs/axe-core/tree/develop/doc
+  */
+
+  getAxeConfiguration() {
+    if (!this.shadowRoot) return null;
+    const script = this.shadowRoot.querySelector<HTMLScriptElement>(
+      '#accessibility-axe-configuration',
+    );
+    if (!script) return null;
+    if (!script.textContent) return null;
+    try {
+      return JSON.parse(script.textContent);
+    } catch (err) {
+      // eslint-disable-next-line no-console
+      console.error('Error loading Axe config');
+      // eslint-disable-next-line no-console
+      console.error(err);
+    }
+
+    // Skip initialization of Axe if config fails to load
+    return null;
+  }
+
+  // Initialise axe accessibility checker
+  async initialiseAxe() {
+    if (!this.shadowRoot) return;
+    const accessibilityTrigger = this.shadowRoot.getElementById(
+      'accessibility-trigger',
+    );
+    if (!accessibilityTrigger) return;
+
+    const config = this.getAxeConfiguration();
+    if (!config) return;
+
+    // Initialise Axe based on the configurable context (whole page body by default) and options ('empty-heading', 'p-as-heading' and 'heading-order' rules by default)
+    const results = (await axe.run(
+      config.context,
+      config.options,
+    )) as unknown as AxeResults;
+
+    const a11yErrorsNumber = results.violations.reduce(
+      (sum, violation) => sum + violation.nodes.length,
+      0,
+    );
+
+    if (results.violations.length) {
+      const a11yErrorBadge = document.createElement('span');
+      a11yErrorBadge.textContent = String(a11yErrorsNumber);
+      a11yErrorBadge.classList.add('w-userbar-axe-count');
+      this.trigger.appendChild(a11yErrorBadge);
+    }
+
+    const dialogTemplates = this.shadowRoot.querySelectorAll(
+      '[data-wagtail-dialog]',
+    );
+    const dialogs = dialog(
+      dialogTemplates,
+      this.shadowRoot as unknown as HTMLElement,
+    );
+    if (!dialogs.length) return;
+
+    const modal = dialogs[0];
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const modalBody = modal.$el.querySelector('[data-dialog-body]');
```

Comment:
If we're putting a ts-ignore directive we should leave a comment explaining why.